lib/puppet/pops/validation/checker4_0.rb

```diff
@@ -535,18 +536,47 @@ def internal_check_future_reserved_word(o, name)
   NO_NAMESPACE = :no_namespace
   NO_PATH = :no_path
   BAD_MODULE_FILE = :bad_module_file
-  def internal_check_file_namespace(o, name, file)
+
+  def internal_check_file_namespace(o)
+    file = o.locator.file
     return if file.nil? || file == '' #e.g. puppet apply -e '...'
 
     file_namespace = namespace_for_file(file)
     return if file_namespace == NO_NAMESPACE
 
     # Downcasing here because check is case-insensitive
-    if file_namespace == BAD_MODULE_FILE || !name.downcase.start_with?(file_namespace)
-      acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    if file_namespace == BAD_MODULE_FILE || !o.name.downcase.start_with?(file_namespace)
+      acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => o.name, :file => file})
     end
   end
 
+  def internal_check_top_construct_in_module(prog)
+    return unless prog.is_a?(Model::Program) && !prog.body.nil?
+
+    #Check that this is a module autoloaded file
+    file = prog.locator.file
+    return if file.nil?
+    return if namespace_for_file(file) == NO_NAMESPACE
+
+    body = prog.body
+    errors = (!body.is_a?(Model::BlockExpression) ? [ valid_top_construct_error(body) ] :
+      body.statements.collect { |s| valid_top_construct_error(s) }).select { |x| !x.nil? }.to_a
+    return if errors.length == 0
+
+    acceptor.accept(Issues::ILLEGAL_TOP_CONSTRUCT_LOCATION, nil, {:errors => errors, file: file})
```

Comment:
Suggest looping over `errors` and calling `acceptor.accept` for each one, giving the AST object to blame instead of `nil` and not passing an array of errors to be formatted.
That way, neither `file` nor `errors` needs to be passed to the issue, and the error can be formatted to say something about the kind of statement that was found (via the issue system's labeling
capability).